### PR TITLE
Memory: Create newtype `ContextId`

### PR DESCRIPTION
**Pull request by plafer**
Creates the `ContextId` type instead of using `u32` everywhere. This makes the code easier to understand to a newcomer; e.g. types such as `(u32, u32, u32)` make understanding the code harder. And this newtype will also prevent bugs which e.g. use the context ID as an address (who's also a `u32`).

I suggest we also introduce a newtype for memory `Address`.

### CODE REVIEW

**Review by bobbinth — COMMENTED** [review]
Thank you! Totally agree with the motivation of the PR (and also think it would be a good idea to have a special type for memory address - though, probably in a different PR). A few preliminary comments:

1. I think a single `ContextId` type will probably suffice. Specifically, I don't think we need another type for `MemoryContextId`. I would also put the definition of `ContextId` into `/src/system/mod.rs` rather than into memory chiplet.
2. I don't think we need to implement arithmetic operations on `ContextId`. As far as I remember, we never actually add/subtract contexts.
3. We usually try to reduce the number of non-dev dependencies as much as possible. So, I would probably manually implement `Display`, `Into` etc. traits on `ContextId` rather than adding an extra dependency.

**Comment by plafer** [comment]
> I don't think we need to implement arithmetic operations on ContextId. As far as I remember, we never actually add/subtract contexts.

The only arithmetic we do is a subtraction [here](https://github.com/0xPolygonMiden/miden-vm/blob/24a8aefa95652818a8b0edc054b6569069160394/processor/src/chiplets/memory/mod.rs#L171). For now I'll implement only `Sub` with a `Sub::Output=u32`, but we can change it

File: processor/src/chiplets/memory/mod.rs [hunk 1]
```diff
@@ -168,7 +170,7 @@ impl Memory {
 
                     // compute delta as difference between context IDs, addresses, or clock cycles
                     let delta = if prev_ctx != ctx {
-                        (ctx - prev_ctx) as u64
+                        (ctx - prev_ctx).into()
```
Comment (bobbinth) [inline processor/src/chiplets/memory/mod.rs:173]: I would first convert the contexts into integers and then perform the subtraction as logically, we should not be able to subtract one context from another.

File: processor/src/decoder/block_stack.rs [hunk 2]
```diff
@@ -1,3 +1,5 @@
+use crate::system::ContextId;
+
 use super::{Felt, Vec, Word, ONE, ZERO};
```
Comment (bobbinth) [inline processor/src/decoder/block_stack.rs:3]: nit: I would remove the blank line between these two imports.

File: processor/src/system/mod.rs [hunk 3]
```diff
@@ -296,3 +301,60 @@ impl System {
         }
     }
 }
+
+// MEMORY CONTEXT
+// ================================================================================================
```
Comment (bobbinth) [inline processor/src/system/mod.rs:306]: nit: I would say "EXECUTION CONTEXT".

File: processor/src/system/mod.rs [hunk 4]
```diff
@@ -296,3 +301,60 @@ impl System {
...
+    }
+}
+
+impl Sub for ContextId {
+    type Output = u32;
+
+    fn sub(self, rhs: Self) -> Self::Output {
+        self.0 - rhs.0
+    }
+}
```
Comment (bobbinth) [inline processor/src/system/mod.rs:354]: I would get rid of this. As mentioned in one of the previous comments, we shouldn't actually be able to subtract contexts. Or at least, subtracting one context from another does not result in a valid context.

File: processor/src/system/mod.rs [hunk 5]
```diff
@@ -2,6 +2,11 @@ use super::{
     ExecutionError, Felt, FieldElement, StarkField, SysTrace, Vec, Word, EMPTY_WORD, ONE, ZERO,
 };
 
+use core::{
```
Comment (bobbinth) [inline processor/src/system/mod.rs:5]: minor nit: I would remove a blank line above this import.

File: processor/src/system/mod.rs [hunk 6]
```diff
@@ -198,7 +203,7 @@ impl System {
     /// for SYSCALLs this remains set to the hash of the last invoked function.
     pub fn start_syscall(&mut self) {
         debug_assert!(!self.in_syscall, "already in syscall");
-        self.ctx = 0;
+        self.ctx = 0.into();
```
Comment (bobbinth) [inline processor/src/system/mod.rs:206]: is there a reason to do `0.into()` rather than `ContextId::root()` here?

**Review by bobbinth — APPROVED** [review]
Thank you! Looks good! I left a few small comments inline. Once these are addressed, we can merge.

File: processor/src/chiplets/memory/mod.rs [hunk 7]
```diff
@@ -1,3 +1,5 @@
+use crate::system::ContextId;
+
 use super::{
```
[the remark this answers is not in the document]
Comment (plafer) [reply processor/src/chiplets/memory/mod.rs]: done

File: processor/src/chiplets/memory/mod.rs (hunk 1 above)
Comment (plafer) [reply processor/src/chiplets/memory/mod.rs]: Done [36e09d1](https://github.com/0xPolygonMiden/miden-vm/pull/1117/commits/36e09d13068176f944fba2907d57557016dc5518)

Although I'd argue that it *needs* to make sense to "take the difference of 2 context IDs", well, because we're doing it here. By converting to `u32` and then subtracting in a way makes it less clear what the type, since with that hack I can also add context IDs, divide them, etc.

Or in other words, if it actually doesn't make sense to subtract context IDs, then is there a way to redesign this code snippet to not rely on it? Could the `delta` here be computed only from the addresses or the clock cycles?

File: processor/src/decoder/block_stack.rs (hunk 2 above)
Comment (plafer) [reply processor/src/decoder/block_stack.rs:3]: done

File: processor/src/system/mod.rs (hunk 3 above)
Comment (plafer) [reply processor/src/system/mod.rs:303]: done

File: processor/src/system/mod.rs (hunk 4 above)
Comment (plafer) [reply processor/src/system/mod.rs]: See [related comment](https://github.com/0xPolygonMiden/miden-vm/pull/1117#discussion_r1370009826)

File: processor/src/system/mod.rs (hunk 5 above)
Comment (plafer) [reply processor/src/system/mod.rs]: done

File: processor/src/system/mod.rs (hunk 6 above)
Comment (plafer) [reply processor/src/system/mod.rs]: Nope; fixed

File: processor/src/chiplets/memory/mod.rs (hunk 1 above)
Comment (plafer) [reply processor/src/chiplets/memory/mod.rs]: Thinking about this some more, it already makes sense to compare `ContextId`s (we derive `PartialOrd` and `Ord`); then why not subtract them? Basically a `ContextId` differs from, say, an `AccountId` in that the root context is the smallest, and every `call()` creates a "greater context ID". The difference between 2 `ContextId`s would be formally defined as the number of `call`s from the root context; which is how we think of the type, and how we use it here.

If you disagree, let's just leave it as is and not spend too much time on this 🙂

Comment (bobbinth) [reply processor/src/chiplets/memory/mod.rs]: > The difference between 2 `ContextId`s would be formally defined as the number of `call`s from the root context

Similar but not exactly like you described. Difference between two contexts is actually the number of clock cycles between these contexts were created. So, root context is created at cycle 0, the next context could be created at cycle 123 and then the difference between them would be 123. The next context could be created as cycles 155, and then the difference between them would be 32. But 32 itself is not a valid context (for execution of this program).

Basically, what we want to do is `ContextId - ContextId = ClockCycleDelta` but I think this probably introduces more complexity than we need. So, let's keep it as is :)

File: processor/src/system/mod.rs [hunk 8]
```diff
@@ -296,3 +297,52 @@ impl System {
         }
     }
 }
+
+// EXECUTION CONTEXT
+// ================================================================================================
+
+/// Represents the ID of a memory context
```
Comment (bobbinth) [inline processor/src/system/mod.rs:304]: minor nit: let's change "memory context" to "execution context" here as well.

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you! I left one small nit inline. After this, let's squash the commits and merge.

File: processor/src/chiplets/memory/mod.rs (hunk 1 above)
Comment (plafer) [reply processor/src/chiplets/memory/mod.rs]: Ah, that makes a lot of sense. Agreed!

File: processor/src/system/mod.rs (hunk 8 above)
Comment (plafer) [reply processor/src/system/mod.rs]: done
